Dshop/apps/products_catalogue/tests/test_api_persistent_cart.py

```diff
@@ -0,0 +1,179 @@
+
+    data = {   
+    'items': [
+            {'product_pk': product.pk, 'quantity': quantity}
+            for product, quantity in zip(ten_tv_products, quantities)
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
+    response = client.post(reverse("api_cart"), {}, format="json")
+    assert response.status_code == status.HTTP_201_CREATED
+    assert_data_empty(response.data)
+
+
+@pytest.mark.django_db
+def test_get_non_unique_pks(tv_product):
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+
+    data = {
+        'items':[
+            {'product_pk': tv_product.pk, 'quantity': 2},
+            {'product_pk': tv_product.pk, 'quantity': 3}
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+@pytest.mark.django_db
+def test_get_non_existing_pks(tv_product):
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    NON_EXISTING_ID = 999999
+        
+    data = {
+        'items':[
+            {'product_pk': tv_product.pk, 'quantity': 2},
+            {'product_pk': NON_EXISTING_ID, 'quantity': 3}
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_404_NOT_FOUND
+    response = client.get(reverse("api_cart"), {}, format="json")
+    assert_data_empty(response.data)
+
+
+@pytest.mark.django_db
+def test_add_check_subtotals(ten_tv_products):
+    three_products = ten_tv_products[:3]
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+    data = {
+        'items':[
+            {'product_pk': three_products[0].pk, 'quantity': 1},
+            {'product_pk': three_products[1].pk, 'quantity': 5},
+            {'product_pk': three_products[2].pk, 'quantity': 10},
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert Decimal(response.data['items'][0]['subtotal']) == three_products[0].price
+    assert Decimal(response.data['items'][1]['subtotal']) == 5 * three_products[1].price
+    assert Decimal(response.data['items'][2]['subtotal']) == 10 * three_products[2].price
+
+@pytest.mark.django_db
+def test_get_zero_quantities(tv_product):
+    client = APIClient()
+    user = User.objects.create_user(username='testuser', password='testpassword')
+    client.force_authenticate(user)
+        
+    data = {
+        'items':[
+            {'product_pk': tv_product.pk, 'quantity': 0}
+        ]
+    }
+    response = client.post(reverse("api_cart"), data, format="json")
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
```

Comment:
Missing get call. 
Overall, those tests are only testing POST method, I didn't notice get requests except for test_get_cart_empty and test_get_non_existing_pks.

Reply:
modified a few post tests to also test get